### PR TITLE
Dac 570 mount public wpsoutputs

### PR DESCRIPTION
**Pull request by ChaamC**
Resolves tasks [DAC-570](https://crim-ca.atlassian.net/browse/DAC-570) and [DAC-571](https://crim-ca.atlassian.net/browse/DAC-571), preparing necessary folders and files for the user access to public and user wpsoutputs data. 
Also, prepares the task [DAC-442](https://crim-ca.atlassian.net/browse/DAC-442), adding an API endpoint for the resync  of a handler and implementing part of the resync operations required for the ``FileSystem`` handler.

Features / Changes : 

* Add monitoring to the ``FileSystem`` handler to watch wps outputs data.
* Synchronize both public and user wps outputs data to the workspace folder with hardlinks for user access.
* Add resync endpoint to trigger a handler's resync operations. Only the ``FileSystem`` handler is implemented for now,
  regenerating hardlinks associated to wps outputs public data.

### CODE REVIEW

File: docs/components.rst
Comment (fmigneault) [inline docs/components.rst:110]: Because all these considerations come from multiple lengthy discussion/reflections, I think it would be worth adding some `.. seealso::` block with references to the Jira issue and other relevant Confluence pages linked in it for traceability. They won't be accessible externally, but they can help up find important decisions quickly if we ever need to come back on this.

File: tests/test_filesystem.py
Comment (fmigneault) [inline tests/test_filesystem.py:31]: Doesn't this get picked up by pytest/unittest as a separate test class (that is executed on its own) even though there are no test cases in it (because of the `Test` prefix). 

If so, consider adding a `Base`, `Util`, or some other prefix. 
This will also highlight the fact that it is intended to be extended by other test classes, and only provides utilities instead of being test cases on their own.

File: tests/test_filesystem.py
Comment (fmigneault) [inline tests/test_filesystem.py:226]: `Path(<file>).touch()` ?

here and other locations

File: tests/utils.py
Comment (fmigneault) [inline tests/utils.py:173]: Use `NotImplementedError` insatead of `pass` +  FIXME comment.

File: tests/utils.py
Comment (fmigneault) [inline tests/utils.py:214]: Idem as other.

File: tests/test_filesystem.py
Comment (fmigneault) [inline tests/test_filesystem.py:1]: That is some nice test suite!  👍

**Review by fmigneault — CHANGES_REQUESTED** [review]
Great job. Really good PR.

File: cowbird/api/handlers/views.py
Comment (ChaamC) [reply cowbird/api/handlers/views.py:51]: It returns a 500 error with the exception "NotImplementedError". 
Ex.: 
```
{"path": "/handlers/Magpie/resync", 
"url": "http://localhost:7000/cowbird/handlers/Magpie/resync", 
"detail": "Internal Server Error. Unhandled exception occurred.", 
"method": "PUT", 
"exception": "NotImplementedError",
 "code": 500, 
"type": "application/json"}
```

The `NotImplementedError` is also displayed in the cowbird logs.

I think it's sufficient, since we have the mentions of the url and the exception?

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:70]: Actually, I think we should just remove ths `get_resource_id` function. It is not used or implemented by any handler right now and I am not sure I see the usefulness of the function. I think it's a remnant of the initial designs.
Any objections to removing this function?

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:249]: Great idea, it will be much cleaner.

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:270]: So, the way I saw the design, only the source wps outputs directory is monitored (e.g.: `/data/wps_outputs/weaver`), aka where the WPS processes write their outputs.

This source directory is synchronized with the 2 destination directories (public and user).

I was considering the destination directories to be used to give access to the WPS outputs, and not necessarily as working directories where the user could add new files manually. That's why I am deleting the file if it exists regardless, considering that the 'truth' is always found in the source directory. So, whatever if there is an existing file already in the destination (being a hardlink or an unrelated file), it would delete it, to add the true source file that should be found there.

I could always have a more forgiving behavior. Displaying a warning in the logs if it's an unrelated file and not deleting it. I can probably detect if it is a hardlink by checking the number of links on the file and also comparing the inode, but if we choose this behavior, we will never delete an existing file anyway.
A disavantage with this design, is that we won't have a guarantee that the files will be valid in the destination folders. They might be unrelated files, or actual WPS output files, but the user won't be able to know, even if we do a resync.

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:283]: As mentioned above, the design I implemented was to only monitor the WPS outputs dir, so new files added to the user workspace dir are ignored.

Is this something that we want to support? If it is a needed use case, I will have to adapt this PR consequently.

File: cowbird/handlers/impl/catalog.py
Comment (fmigneault) [reply cowbird/handlers/impl/catalog.py:79]: Since it is planned to be worked on later and that the response is descriptive enough, the FIXME by itlsef is ok.

File: cowbird/handlers/impl/filesystem.py
Comment (fmigneault) [reply cowbird/handlers/impl/filesystem.py:70]: I thought this was "implementing" an abstract method from the base handler. 
If it's not the case, then it is indeed better to remove it altogether.

File: cowbird/handlers/impl/filesystem.py
Comment (fmigneault) [reply cowbird/handlers/impl/filesystem.py:270]: > That's why I am deleting the file if it exists regardless, considering that the 'truth' is always found in the source directory.

I agree with this. Users should not have the possibility to add more files to an already executed WPS process outputs.
The best would be to actually block them with read-only directories/files, but if this cannot be achieved, the unidirectional sync from WPS outputs as "truth" is good enough for now.

Given that the user that could write additional files within its user workspace would not see the log output, I don't think it would add any value. It's good to keep it as you currently implemented it.

File: cowbird/handlers/impl/filesystem.py
Comment (fmigneault) [reply cowbird/handlers/impl/filesystem.py:283]: If we can block it, it's better, but not mandatory. We don't want to support push files back to the WPS outputs from the user workspace.

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:70]: It is indeed implementing an abstract method from the base handler, but no handler actual implement it (only raises `NotImplementedError`). I don't think there is any current or future usefulness to this method, so I will remove it.

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:270]: The thing is the `secure-data-proxy` is adding the option of having `read` and/or `write` permissions on the user WPS output files. If we want to support `write` permissions on those files, I tested it on JupyterLab and I need to also add `write` permissions to the parent folder too, so that the file can be modified by the user via the JupyterLab interface. I didn't work if I added only the `write` permissions on the file only. Note that those folders found in the user workspace are not linked to those found in the WPS outputs source folder, so the folders don't have to use corresponding permissions with those found from the source.

So, currently, by default, all user hardlinks are `read`/`write` if the service `secure-data-proxy` doesn't exist, or, if it exists,  the permissions are resolved according to the `secure-data-proxy`'s permissions. For now, I was making the parent directory of a user hardlink `read`/`write` in every case to allow possible `write` permissions on any of their potential children files and to simplify the process, considering that the children files could have a mix of different permissions on them.

It is possible to make the user hardlinks and parent folders read-only (like the public hardlinks), but I'm not sure if it is what we want?

So I see these options :
- Keep the current design, which would make it possible for the users to create files in user WPS folder. These created files, which would be unrelated to the source WPS folder, could disappear without warning to the user if we resync the handler. In this case, we could probably add a mention about this in the eventual README (see [DAC-545](https://crim-ca.atlassian.net/browse/DAC-545)?)
- Ignore the `secure-data-proxy` entirely and making everything read-only, like the public WPS hardlinks
- Keep checking for the `secure-data-proxy`, but only use read permissions on the folders and hardlinks, ignoring the `write` permissions of the `secure-data-proxy`. This would prevent the user from creating unrelated files in its user WPS output folder.

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:45]: I had made it that way, because I saw the variable `wps_outputs_public_subdir` as optional. The other explicit params are required in the `config.yml` file and `kwargs` params can be optionally used in the `config.yml`.

The reason I saw this parameter as more optional, is that it is only managed by the cowbird service and I saw less necessity of it being customized for different setups. It is for a directory that is managed by the Cowbird service only, as opposed to for example `jupyterhub_user_data_dir` and `wps_outputs_dir` which are used by other services (JupyterHub, Weaver, and maybe others).

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:321]: As mentionned in earlier responses, the destination folders containing the hardlinks are assumed not to contain anything important since it should only contain hardlinks to files from the source WPS outputs dir. Destination folders should not be used to stored anything else, so here we delete the path regardless of the content.

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:319]: Changed to `shutil.rmtree()`

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:403]: As other responses, hardlinks are removed regardless, considering the destination folders to contain only copies of the source dirs and not unrelated files.

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:434]: Renamed get methods to `get_user_workspace_wps_outputs_dir` and `get_public_workspace_wps_outputs_dir`, and adjusted related variables consequently.

File: cowbird/monitoring/monitoring.py
Comment (ChaamC) [reply cowbird/monitoring/monitoring.py:55]: I think it is not needed as it will be done in the `get_handler()` method, returning None, if the handler is not defined in the config or if it has `active=False`.

See [here](https://github.com/Ouranosinc/cowbird/blob/DAC-570-mount-public-wpsoutputs/cowbird/handlers/handler_factory.py#L78-L82)

File: tests/utils.py
Comment (ChaamC) [reply tests/utils.py:173]: I used `pass` specifically here (and the other mentioned lined below) instead of `NotImplementedError`, since this is just a Mocking class. This function might never be implemented even if the actual method found in the real `Magpie` class gets eventually implemented.

This is similar to the rest of the Mock class. Other functions such as `user_created`, `user_deleted`, etc. are not implemented and only have a pass, since they don't need to have any content for the current tests.

I would keep the `pass`, as I don't see a reason to raise an exception here.

File: tests/utils.py
Comment (ChaamC) [reply tests/utils.py:214]: See other response.

File: tests/test_filesystem.py
Comment (ChaamC) [reply tests/test_filesystem.py:31]: If I run the tests, I don't necessarily see tests from this class, as it doesn't contain any.

I think I could add a `Base` prefix to make it more clear that it doesn't contain any tests, but I think unittest still detects it as a test class since it inherits `unittest.TestCase`. I need this base class to inherit `unittest.TestCase` though since I want this class to execute the setup and teardown functions during the tests of the children classes.

File: docs/components.rst
Comment (ChaamC) [reply docs/components.rst:76]: Do you mean to add a mention that permissions should not be handled via the FileSystem and only from Magpie? That if a permission is changed directly on the FileSystem, it will be ignored?

Comment (ChaamC) [reply docs/components.rst:76]: I will add a mention about this, because it's worth mentioning anyway, but let me know if you meant something else.

File: tests/test_filesystem.py
Comment (fmigneault) [reply tests/test_filesystem.py]: Ok. As long as it doesn't produce a +1 test entry, that's fine. Even if it did, not a deal-breaker either.

File: tests/utils.py
Comment (fmigneault) [reply tests/utils.py:170]: I see it more as a safeguard. If the mock method gets called by mistake somehow, the explicitly raised `NotImplementedError` will indicate it clearly. The message of that raised error could also indicate something such as "This mock method implementation should not be called" to make the intention even more verbose.

**Review by fmigneault — COMMENTED** [review]
Just need to review the last comments.

File: tests/utils.py
Comment (ChaamC) [reply tests/utils.py:170]: I get your idea, but I'm not sure if it applies nicely here. I works if I add `NotImplementedError` to the `resync` method only, but there are other methods in the Mock class that use the same `pass` behavior. I think it wouldn't be coherent to use different behaviors (`pass` or `NotImplementedError`) for similar methods, which are either implemented, not implemented yet or will never be implemented on the actual `Handler` classes.

Also, if they all use the `NotImplementedError`, I think it will break multiple tests that use Mock classes. Some tests call API endpoints, which can use Mock handlers, and which would now break, having errors during the API call. Other tests use Mock handlers for testing celery request tasks.

I could always try to change those tests, but I am not sure if it is worth the effort?

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:281]: I agree that blocking the file creation on the user workspace would be ideal, but not possible with the usage of `write` permissions with the `secure-data-proxy` (see my other [comment](https://github.com/Ouranosinc/cowbird/pull/40#discussion_r1297145422) for the longer explanation). 
So I will keep the implementation as it is for now, ignoring new files added to the user workspace wps outputs directory.

**Comment by ChaamC** [comment]
I updated the variables for the `FileSystem` handlers, as we had discussed having configurable options.

For the rest, I should have responded to all the remaining comments. 
Let me know if further adjustments are required or not!

Also, there is a new error on the PR for the `readthedocs` step. It mentions :
```
Problem in your project's configuration. Invalid configuration option "python.system_packages". 
```

I'm not sure where to find that option, is this something on the repo that only admins have access to?

**Comment by fmigneault** [comment]
@ChaamC 
The error about RDT is already fixed in master.

File: cowbird/handlers/impl/filesystem.py
Comment (fmigneault) [reply cowbird/handlers/impl/filesystem.py:46]: Adding it as an actual kw parameter will make it such that you don't need to do `kwargs.get("<kw>", None)` for each of those variables that can be passed as input anyway, and it would make the class slightly better documented. 

You can still leave them optional with `wps_outputs_public_subdir: Optional[str] = None`.

**Review by fmigneault — COMMENTED** [review]
Good to go. 
If you can slide in the tweak about `kwargs` before merge, it would be best.

File: tests/utils.py
Comment (fmigneault) [reply tests/utils.py:170]: In places that use `pass` as a "don't care" or some kind of override for testing purposes, it is ok to leave it as is to skip the resync that "should" normally be called. 

The issue was more for places that use `pass` instead of `NotImplementedError` for cases that are "pending implementation".

In this case, seems to be ok.

File: cowbird/handlers/impl/filesystem.py
Comment (fmigneault) [reply cowbird/handlers/impl/filesystem.py:49]: I didn't think about passing the `DEFAULT_...` values directly, but that's an even better idea! 
Given that, the type would actually be `str` directly. 
Is schema validation performed to ensure those parameters can be `str` only. 
Otherwise, a config that explicitly sets `notebooks_dir_name: null` could technically pass a `None` as parameter.

**Review by fmigneault — APPROVED** [review]
See my last comment. 
If the types are validated, then could to merge.

File: cowbird/handlers/impl/filesystem.py
Comment (fmigneault) [reply cowbird/handlers/impl/filesystem.py:282]: I think it is a good idea to have some details in a README about automatic removal of the files due to how sync could behave, such that the user can be informed of why his files could disappear, and why it is not recommended to add any more files in there. 

I'm not too worried about the `write` on the directory, since the unidirectional sync will not propagate back to the source WPS outputs. However, doesn't `write` on the hardlink allow the user to modify the original file? For example, doing something like `cat "random" >> "<hardlink/file>"` could modify the source WPS output? If this is the case, then it is something that should definitely be prevented with `read` only access using the permissions of `secure-data-proxy`.

File: cowbird/handlers/impl/filesystem.py
Comment (fmigneault) [reply cowbird/handlers/impl/filesystem.py]: Technically, `SECURE_DATA_PROXY_NAME` should not be needed if the component is not enabled in the stack, but I guess this is not something that matters a lot because the specific name is only relevant and used when exactly the same service is found on Magpie. 

I think they could both be optional in the config, but they must resolve to a "required"/defined `str` using those defaults for the code to work.

Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py]: I will put it as optional, with a default value if None.

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:49]: I will update the schema validation, which didn't check for those new optional parameters, also added a restriction that the string should not be empty, to prevent weird cases.

File: cowbird/handlers/impl/filesystem.py
Comment (ChaamC) [reply cowbird/handlers/impl/filesystem.py:282]: Yes, indeed. The `write` on the hardlink does modify the original source file, and I was aware of that. I assumed it was okay, since it is a user-specific file, and the `secure-data-proxy` is allowing specification of `write` permissions on the file. 
I can adjust it if we want to change this behavior. Should I ignore `write` permissions on `secure-data-proxy` on Magpie though? It seems it might lead to confusion if a `write` permissions is added but the file stays read-only.

But, yes I can change it. I could only check the `read` permissions on `secure-data-proxy` to adjust the file permissions, and I could assign `read-only` on the hardlinks by default if the `secure-data-proxy` service doesn't exist, ignoring all `write` permissions.

Let me know if we should go ahead with this design, or if you have any idea about how to avoid confusion with the `write` permissions on Magpie (except to document it in .rst files).

File: docs/conf.py
Comment (fmigneault) [inline docs/conf.py:153]: See if  https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-linkcheck_anchors_ignore works instead. It would be a better ignore condition.

File: cowbird/handlers/impl/filesystem.py
Comment (fmigneault) [reply cowbird/handlers/impl/filesystem.py:280]: Yes. It should be modified, but don't ignore the permission in Cowbird code. If `write` is explicitly set, then it would be valid to let override of the file by the user. Cowbird would sync the permission normally. This would make sense in the case of `secure-data-proxy` for other directories than WPS outputs. However, WPS outputs themselves should never have `write`, since they are produced by some processes and their result should remain constant. If `write` was set somewhere for WPS output directories (whether for `{user}` or `public`), that would be an error (ie: bad Cowbird *configuration* or some improperly defined Magpie hook).

File: docs/conf.py
Comment (ChaamC) [reply docs/conf.py]: It works but it only uses the anchor part, so I have to put the failing links specific anchors. I updated the code accordingly.

Note : I saw there was also the [linkcheck_anchors_ignore_for_url](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-linkcheck_anchors_ignore_for_url) option, but the `sphinx-build` version on the project is not recent enough to have this feature.

**Comment by ChaamC** [comment]
@fmigneault Updated according to latest feedback. Removed usage of write permissions on WPS outputs data.

File: docs/components.rst
Comment (fmigneault) [inline docs/components.rst:82]: All ``` `secure-data-proxy` ``` should be with double ``` `` ```, otherwise they appear as plain text (see https://github.com/Ouranosinc/cowbird/blob/5595572ce915c6f66ca4dcf51dd4bb07986e0786/docs/components.rst).

File: tests/test_filesystem.py
Comment (fmigneault) [inline tests/test_filesystem.py:373]: Can you add a ``.. note::`` with a very brief summary and a reference to the doc where ``others`` permission is explained instead of using ``user``/``group``? Reading the tests without this context feels odd due to the `660`/`664` permissions. 

Also add a 
```rst
.. seealso::
    - :func:`cowbird.utils.update_filesystem_permissions` 
```

And maybe you need to adjust its description. It mentions "(user only)" although it is working on `others` permissions. 
The function itself could also reference the docs for the detailed explanation.

File: tests/test_filesystem.py
Comment (fmigneault) [inline tests/test_filesystem.py:509]: Add some note next/above this new test case (and other below) indicating that `write` should be ignored by design, which is why `0o666` is not the expected value. Otherwise, it doesn't make sense in comparison with the other test cases why this value is expected.

**Review by fmigneault — COMMENTED** [review]
Looks all good implementation wise. Just a few notes for wrapping up the documentation.

**Comment by ChaamC** [comment]
@fmigneault 
I see in the birdhouse-deploy repo that both paths `/data/wpsoutputs` and `/data/wps_outputs` are used. Is there some kind of standard we should decide to use, either with or without the underscore?
This could affect some variables I'm using in this PR too.

**Comment by fmigneault** [comment]
> I see in the birdhouse-deploy repo that both paths `/data/wpsoutputs` and `/data/wps_outputs` are used. Is there some kind of standard we should decide to use, either with or without the underscore?

I believe only the nginx endpoint uses `/wpsoutputs`, but all directories normally use `/data/wps_outputs`.
Where did you see `/data/wpsoutputs`? That is probably a mistake.

File: docs/conf.py
Comment (fmigneault) [reply docs/conf.py]: Ok. You can try using it with a more recent version of Sphinx and see if it works. I think the `<7` requirement was only set to avoid flaky build because v7 was recent at that time. Using 7.1 is probably ok. If the build fails or causes more trouble that it is worth, then your current edit is fine as well.

**Comment by ChaamC** [comment]
> > I see in the birdhouse-deploy repo that both paths `/data/wpsoutputs` and `/data/wps_outputs` are used. Is there some kind of standard we should decide to use, either with or without the underscore?
> 
> I believe only the nginx endpoint uses `/wpsoutputs`, but all directories normally use `/data/wps_outputs`. Where did you see `/data/wpsoutputs`? That is probably a mistake.

A search of `/data/wpsoutputs` in the `birdhouse-deploy` repo brings multiple matches, for example in multiple bird configs (generic_bird, emu, raven, finch, flyingpigeon, ...)

Should I change every values or variable/function names to use a particular convention, or is it okay as it is? Currently, I used the `wpsoutputs` format for paths or resource name, but also used `wps_outputs` in other places, since it wasn't clear to me what was the convention on birdhouse.

**Comment by fmigneault** [comment]
> A search of `/data/wpsoutputs` in the `birdhouse-deploy` repo brings multiple matches, for example in multiple bird configs (generic_bird, emu, raven, finch, flyingpigeon, ...)

Yes, but those are mounted paths with resolve to `/data/wps_outputs`:
```yml
services:
  finch:
    volumes:
      - wps_outputs:/data/wpsoutputs
volumes:
  wps_outputs:
    driver_opts:
      type: none
      device: /data/wps_outputs
      o: bind
```
When these paths are mounted in Cowbird, they should use the locations relevant to Cowbird volumes, or use hardlinks as defined on the server side.

**Comment by fmigneault** [comment]
@ChaamC 
Need to merge https://github.com/Ouranosinc/cowbird/commit/0ad27b938ded22537d444889e681cc9320b2473e (master) to fix the RTD error seen on https://readthedocs.org/projects/pavics-cowbird/builds/21935569/

**Comment by ChaamC** [comment]
@fmigneault 
Ok, I will do the master merge and will update variables to use the `wps_outputs` format.

**Review by fmigneault — APPROVED** [review]
Great job. Thanks for all the work!

**Pull request merged**
